src/hotspot/os/linux/os_linux.cpp

```diff
@@ -3753,8 +3714,60 @@ void os::large_page_init() {
     return;
   }
 
-  size_t large_page_size = Linux::setup_large_page_size();
-  UseLargePages          = Linux::setup_large_page_type(large_page_size);
+  // 2) Scan OS info
+  size_t default_large_page_size = scan_default_large_page_size();
+  os::Linux::_default_large_page_size = default_large_page_size;
+  if (default_large_page_size == 0) {
+    // No large pages configured, return.
+    UseTransparentHugePages = false;
+    UseHugeTLBFS = false;
+    UseSHM = false;
+    return;
+  }
+  os::PageSizes all_large_pages = scan_multiple_page_support();
+
+  // 3) Consistency check and post-processing
+
+  // It is unclear if /sys/kernel/mm/hugepages/ and /proc/meminfo could disagree. Manually
+  // re-add the default page size to the list of page sizes to be sure.
+  all_large_pages.add(default_large_page_size);
+
+  // Populate large page sizes to _page_sizes. Add pages that
+  // are less than or equal to LargePageSizeInBytes, except when LargePageSizeInBytes=0
+  // or FLAG_IS_DEFAULT(LargePageSizeInBytes), add all sizes
+  for (size_t page_size = all_large_pages.largest(); page_size != 0;
+         page_size = all_large_pages.next_smaller(page_size)) {
+    if (LargePageSizeInBytes != 0
+        || !FLAG_IS_DEFAULT(LargePageSizeInBytes)) {
+      if (page_size <= LargePageSizeInBytes){
+        _page_sizes.add(page_size);
+      }
+    } else {
+      _page_sizes.add(page_size);
+    }
+  }
+
+  LogTarget(Info, pagesize) lt;
+  if (lt.is_enabled()) {
+    LogStream ls(lt);
+    ls.print("Available page sizes: ");
+    _page_sizes.print_on(&ls);
+  }
+
+  // Handle LargePageSizeInBytes
+  if (!FLAG_IS_DEFAULT(LargePageSizeInBytes) && LargePageSizeInBytes != default_large_page_size) {
+    if (all_large_pages.contains(LargePageSizeInBytes)) {
+      _large_page_size = LargePageSizeInBytes;
+      log_info(os)("Overriding default large page size using LargePageSizeInBytes");
+    } else {
+      _large_page_size = default_large_page_size;
+    }
+  } else {
+    _large_page_size = default_large_page_size;
+  }
```

Comment:
Move this part up to be done before you populate `_page_sizes` with the page sizes from `all_page_sizes`. Then you can use `_large_page_size` as the maximum page size that you add to `_page_sizes` and you don't have to care about `LargePageSizeInBytes` in that for-loop.

Reply:
I also think that the log-tag should be `pagesize`.

Reply:
The comment here could also be updated: 
```
// Check LargePageSizeInBytes and setup _large_page_size to hold 
// the maximum allowed large page size. If not set the default large 
// page size is used as the maximum. ```

Reply:
+1